### ISSUE TITLE
Support for aarch64/ARMv8/ARM64 architecture with Docker

### ISSUE THREAD

**Issue by mirobertod** [localstack/localstack#2550]
# Type of request: This is a ...
[ ] bug report
[X ] feature request

# Detailed description
Currently, on [docker hub](https://hub.docker.com/r/localstack/localstack/tags) the only architecture supported is `linux/amd64`.
I think ARM architecture will be more implemented in the next future, both in consumer and business environment.
Nowadays, we can see most of the open-source software have ARM binaries and cloud computing companies like AWS are [investing in this architecture](https://aws.amazon.com/blogs/aws/new-m6g-ec2-instances-powered-by-arm-based-aws-graviton2/).

## Expected behavior
I'll be glad if it would be possibile to have the Docker image compatible with ARM architecture.

## Actual behavior
No docker image for ARM architecture.

Thank you

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-199) by [Unito](https://www.unito.io/learn-more)

**Comment by whummer** [comment]
Thanks for this suggestion @mirobertod . Not sure if supporting ARM is a realistic option at this point, given our set of dependencies, but would be great to get an initial understanding of the complexity of such an undertaking. Do you think you could give it a try and create some initial experiments to see if ARM can be realistically supported in the future? That would be highly appreciated - thanks!

**Comment by mirobertod** [comment]
Hi @whummer 
I can give it a try. I see that the `Dockerfile` of the localstack image starts with `FROM localstack/java-maven-node-python` 

Where can I find that `Dockerfile`? [Is it this one?](https://github.com/localstack/localstack/blob/master/bin/Dockerfile.base)
I'm trying to build an ARM compatible Docker image.

**Comment by mirobertod** [comment]
Hi @whummer 
I've been working with the ARM build.
I have some problem with DynamoDB. I see you are using a custom version of it. ([link](https://github.com/localstack/localstack/blob/master/bin/Dockerfile.base#L67))

Your version doesn't include the `libsqlite4java-linux-aarch64.so` which is included in the [official release](https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip) from AWS.
I tried to copy just the `.so` library into your archive but then I get `SIGSEGV JVM`, can you give me a clue how you built the custom version of DynamoDB? Is there a way to add the ARM binary for the library `libsqlite4java`?

Thanks

**Comment by natdan** [comment]
@mirobertod Any luck so far? Have you shared somewhere how far have you got? Anything I can help with?

**Comment by mirobertod** [comment]
> @mirobertod Any luck so far? Have you shared somewhere how far have you got? Anything I can help with?

No luck so far. I tried to build the Docker image on ARM platform (changing the relative binaries from x64 to aarch64). You can read in the previous comment where I am stuck.

I haven't tried again since that time.

**Comment by natdan** [comment]
I suspect you didn't fork this repo, so could you share your Dockerfile? (or fork it and commit it there :) )
I might be able to invest some time and try to understand what originally was made in custom version of DynamoDB (https://github.com/whummer/dynamodb-local/raw/master/etc/DynamoDBLocal.zip) and maybe recreate it with all arm (aarch64) libraries left in place. Unless we get some tips from @whummer :)

**Comment by mirobertod** [comment]
@natdan You are right, I didn't. I can attach my diff here (it's very small) 

```

diff --git a/Dockerfile b/Dockerfile
index c9aeaa2..9fcde9f 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,4 +1,4 @@
-FROM localstack/java-maven-node-python
+FROM localstack-base:arm-2
 
 MAINTAINER Waldemar Hummer (waldemar.hummer@gmail.com)
 LABEL authors="LocalStack Contributors"
diff --git a/bin/Dockerfile.base b/bin/Dockerfile.base
index 5eccc3e..d0263c9 100644
--- a/bin/Dockerfile.base
+++ b/bin/Dockerfile.base
@@ -10,8 +10,8 @@ RUN apk add --no-cache autoconf automake build-base ca-certificates curl cyrus-s
     update-ca-certificates
 
 # install Docker (CLI only)
-RUN docker_version=17.05.0-ce; \
-  curl -fsSLO https://get.docker.com/builds/Linux/x86_64/docker-$docker_version.tgz \
+RUN docker_version=17.09.0-ce; \
+  curl -fsSLO https://download.docker.com/linux/static/stable/aarch64/docker-$docker_version.tgz \
   && tar xzvf docker-$docker_version.tgz \
   && mv docker/docker /usr/local/bin \
   && rm -r docker docker-$docker_version.tgz
@@ -63,15 +63,15 @@ RUN ln -s /usr/bin/python3 /usr/bin/python; ln -s /usr/bin/pip3 /usr/bin/pip
 RUN pip install supervisor
 
 # init environment and cache some dependencies
-# ARG DYNAMODB_ZIP_URL=https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip
+ARG DYNAMODB_ZIP_URL=https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip
 # TODO: Temporarily using a fixed version of DDB in Alpine, as we're hitting a SIGSEGV JVM crash with latest
-ARG DYNAMODB_ZIP_URL=https://github.com/whummer/dynamodb-local/raw/master/etc/DynamoDBLocal.zip
+# ARG DYNAMODB_ZIP_URL=https://github.com/whummer/dynamodb-local/raw/master/etc/DynamoDBLocal.zip
 RUN mkdir -p /opt/code/localstack/localstack/infra && \
     mkdir -p /opt/code/localstack/localstack/infra/dynamodb && \
       curl -L -o /tmp/localstack.ddb.zip ${DYNAMODB_ZIP_URL} && \
       (cd localstack/infra/dynamodb && unzip -q /tmp/localstack.ddb.zip && rm /tmp/localstack.ddb.zip) && \
     curl -L -o /tmp/localstack.es.tar.gz \
-        https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.1.0-linux-x86_64.tar.gz && \
+        https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.8.0-linux-aarch64.tar.gz && \
     curl -L -o /tmp/elasticmq-server.jar \
         https://s3-eu-west-1.amazonaws.com/softwaremill-public/elasticmq-server-0.15.7.jar && \
     (cd localstack/infra/ && tar -xf /tmp/localstack.es.tar.gz && \
diff --git a/localstack/constants.py b/localstack/constants.py
index f77b817..a6c8f4f 100644
--- a/localstack/constants.py
+++ b/localstack/constants.py
@@ -73,6 +73,7 @@ LAMBDA_TEST_ROLE = 'arn:aws:iam::%s:role/lambda-test-role' % TEST_AWS_ACCOUNT_ID
 
 # installation constants
 ELASTICSEARCH_URLS = {
+    '7.8.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.8.0-linux-aarch64.tar.gz',
     '7.4.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.4.0-linux-x86_64.tar.gz',
     '7.1.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.1.0-linux-x86_64.tar.gz',
     '6.7.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-6.7.0.zip'

```

**Comment by natdan** [comment]
@ mirobertod  Thanks. I'll give it a go. Yesterday I tried to build one by just changing Docker.base to 
```
FROM arm64v8/node:alpine
```
But I haven't got far enough as I did something nasty to my Kubernetes cluster and spend rest of the day rebuilding it...

BTW In the mean time I checked what  zip file from https://github.com/whummer/dynamodb-local/raw/master/etc/DynamoDBLocal.zip is like and it seems that it actually includes `libsqlite4java-linux-aarch64.so` by default - script to create it just adds arm64.so version as well...

**Comment by natdan** [comment]
Finally got some time to play with it (read whole yesterday morning and last night) and here's what I got to:

`com.almworks.sqlite4java.SQLiteException: [-91] cannot load library: java.lang.UnsatisfiedLinkError: /opt/code/localstack/localstack/infra/dynamodb/DynamoDBLocal_lib/libsqlite4java-linux-aarch64.so: **Error loading shared library ld-linux-aarch64.so.1**: No such file or directory (needed by /opt/code/localstack/localstack/infra/dynamodb/DynamoDBLocal_lib/libsqlite4java-linux-aarch64.so)
`

It seems to me that the issue is in base docker image and not version of dynamodb code (although this is with your patch). I'll play more with it...

**Comment by natdan** [comment]
Yes, that seems to be it. I've put 
```
RUN apk add  libc6-compat
```
to the top `bin/Dockerfile.base` and it cured it. At least version with @mirobertod 's patch.

**Comment by natdan** [comment]
Since I'm already spamming the chat here, here's the winning patch I got to successfully make both x86_64 and arm64 images:
```
diff --git a/bin/Dockerfile.base b/bin/Dockerfile.base
index f4c2775..40bf22c 100644
--- a/bin/Dockerfile.base
+++ b/bin/Dockerfile.base
@@ -3,6 +3,8 @@ FROM node:alpine
 MAINTAINER Waldemar Hummer (waldemar.hummer@gmail.com)
 LABEL authors="Waldemar Hummer (waldemar.hummer@gmail.com)"
 
+RUN ([[ `uname -m` == 'aarch64' ]] && apk add libc6-compat) || true
+
 # install some common libs
 RUN apk add --no-cache autoconf automake boost-dev build-base ca-certificates curl cyrus-sasl-dev g++ gcc \
         git groff iputils libffi-dev libtool linux-headers make openssl openssl-dev postgresql-dev \
@@ -10,8 +12,11 @@ RUN apk add --no-cache autoconf automake boost-dev build-base ca-certificates cu
     update-ca-certificates
 
 # install Docker (CLI only)
-RUN docker_version=17.05.0-ce; \
-  curl -fsSLO https://get.docker.com/builds/Linux/x86_64/docker-$docker_version.tgz \
+RUN [[ `uname -m` == 'aarch64' ]] && docker_version=17.09.0-ce || docker_version=17.05.0-ce; \
+  [[ `uname -m` == 'aarch64' ]] && DOCKER_URL=https://download.docker.com/linux/static/stable/aarch64/docker-$docker_version.tgz \
+    || DOCKER_URL=https://get.docker.com/builds/Linux/x86_64/docker-$docker_version.tgz; \
+  echo Downloading $DOCKER_URL; \
+  curl -fsSLO $DOCKER_URL \
   && tar xzvf docker-$docker_version.tgz \
   && mv docker/docker /usr/local/bin \
   && rm -r docker docker-$docker_version.tgz
@@ -66,12 +71,13 @@ RUN pip install supervisor
 # ARG DYNAMODB_ZIP_URL=https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip
 # TODO: Temporarily using a fixed version of DDB in Alpine, as we're hitting a SIGSEGV JVM crash with latest
 ARG DYNAMODB_ZIP_URL=https://github.com/whummer/dynamodb-local/raw/master/etc/DynamoDBLocal.zip
-RUN mkdir -p /opt/code/localstack/localstack/infra && \
+RUN [[ `uname -m` == 'aarch64' ]] && ELASTICSEARCH_URL=https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.8.0-linux-aarch64.tar.gz \
+     || ELASTICSEARCH_URL=https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.1.0-linux-x86_64.tar.gz; \
+    mkdir -p /opt/code/localstack/localstack/infra && \
     mkdir -p /opt/code/localstack/localstack/infra/dynamodb && \
       curl -L -o /tmp/localstack.ddb.zip ${DYNAMODB_ZIP_URL} && \
       (cd localstack/infra/dynamodb && unzip -q /tmp/localstack.ddb.zip && rm /tmp/localstack.ddb.zip) && \
-    curl -L -o /tmp/localstack.es.tar.gz \
-        https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.1.0-linux-x86_64.tar.gz && \
+    curl -L -o /tmp/localstack.es.tar.gz ${ELASTICSEARCH_URL} && \
     curl -L -o /tmp/elasticmq-server.jar \
         https://s3-eu-west-1.amazonaws.com/softwaremill-public/elasticmq-server-0.15.7.jar && \
     (cd localstack/infra/ && tar -xf /tmp/localstack.es.tar.gz && \
diff --git a/localstack/constants.py b/localstack/constants.py
old mode 100644
new mode 100755
index 67c7b41..d1fe80f
--- a/localstack/constants.py
+++ b/localstack/constants.py
@@ -73,6 +73,7 @@ LAMBDA_TEST_ROLE = 'arn:aws:iam::%s:role/lambda-test-role' % TEST_AWS_ACCOUNT_ID
 
 # installation constants
 ELASTICSEARCH_URLS = {
+    '7.8.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.8.0-linux-aarch64.tar.gz',
     '7.4.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.4.0-linux-x86_64.tar.gz',
     '7.1.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.1.0-linux-x86_64.tar.gz',
     '6.7.0': 'https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-6.7.0.zip'
```

@whummer Is it a worth thinking making a PR for it? (all suggestions for improving it are more than welcome and I'm happy to try it out)

**Comment by natdan** [comment]
Actually - I was wrong :( - dynamodb works only in 'inMemory' mode - not when DATA_DIR is set. And that has nothing to do with localstack but dynamodb itself. I've downloaded separate, the latest version and just run in directly on docker instance - and it still didn't work. There might be something else missing from the setup (aside of `ld-linux-aarch64.so.1`) - but so far I've observed three different behavior:
- java dump (something falling in `libsqlite4java-linux-aarch64.so` (*)) - after making  zero sized `/tmp/localstack/data/dynamodb/shared-local-instance.db`
- java process just hangs without making db file

(*) - stack trace logged in dump is:
```
Stack: [0x0000007f7a389000,0x0000007f7a589ab0],  sp=0x0000007f7a588580,  free space=2045k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  0x000000000000dc50
C  [libsqlite4java-linux-aarch64.so+0x54bc0]  pager_write+0x1e8
C  [libsqlite4java-linux-aarch64.so+0x5ab54]  newDatabase.part.539+0x24
C  [libsqlite4java-linux-aarch64.so+0x72038]  sqlite3BtreeBeginTrans+0x410
C  [libsqlite4java-linux-aarch64.so+0x856e4]  sqlite3VdbeExec+0x3e3c
C  [libsqlite4java-linux-aarch64.so+0x8a67c]  sqlite3_step+0x3c4
j  com.almworks.sqlite4java._SQLiteSwiggedJNI.sqlite3_step(J)I+0
j  com.almworks.sqlite4java._SQLiteSwigged.sqlite3_step(Lcom/almworks/sqlite4java/SWIGTYPE_p_sqlite3_stmt;)I+4
j  com.almworks.sqlite4java.SQLiteStatement.step()Z+50
j  com.amazonaws.services.dynamodbv2.local.shared.access.sqlite.SQLiteDBAccess$1.doWork()Ljava/lang/Long;+96
```

**Comment by rpvilao** [comment]
Hi,

Is the solution being merged at some point? :) I tried the patch with no success, maybe the source already deviated too much?

Thanks!

**Comment by ryanone** [comment]
@natdan Just curious, what kind of ARM machine/device are you trying run DynamoDB with? Before trying to modify the Dockerfile, I wanted to see if I were to experience any issues with just [running DynamoDB locally](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.DownloadingAndRunning.html), and for me, I was able to get it successfully running.

**Comment by ryanone** [comment]
I got this working now on my aarch64 instance, although I had to do some additional stuff:

- Install a couple of Linux packages: `postgresql` and `libpq-dev`

- Install a couple of Python packages: `psycopg2-binary` and `pyjnius`
- Build an ARM-compatible binary of local-kms. I tweaked the localstack KMS installer to use that binary, and saw that it worked. I plan on reaching out to the local-kms project owner to see if he's willing to host an ARM-compatible binary.
- I created a Dockerfile based off the current one and made an aarch64-specific version. I then updated the makefile to check the CPU architecture and select between the ARM file or the x86 (current base) file.

After that, it all worked (see below).

@whummer If you're ok with accepting a PR for ARM support, I see the following as the action items:

1. local-kms - Ask the project owner if he wants to host an ARM binary, or generate one and host it somewhere.
2. Modify the [KMS installer](https://github.com/localstack/localstack/blob/master/localstack/services/install.py#L148) based on no. 1 above
3. Make changes to the Dockerfile and Makefile to support aarch64

```
$ make infra
(. .venv/bin/activate; exec bin/localstack start --host)
Starting local dev environment. CTRL-C to quit.
LocalStack version: 0.12.3
2020-12-21T18:12:52:INFO:localstack.utils.analytics.profiler: Execution of "load_plugin_from_path" took 7683.562994003296ms
2020-12-21T18:12:52:INFO:localstack.utils.analytics.profiler: Execution of "load_plugins" took 7684.215068817139ms
Starting edge router (https port 4566)...
Starting mock ACM service on http port 4566 ...
2020-12-21T18:12:53:INFO:localstack.multiserver: Starting multi API server process on port 53295
[2020-12-21 18:12:53 -0800] [35248] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
2020-12-21T18:12:53:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
[2020-12-21 18:12:53 -0800] [35248] [INFO] Running on http://0.0.0.0:53295 (CTRL + C to quit)
2020-12-21T18:12:53:INFO:hypercorn.error: Running on http://0.0.0.0:53295 (CTRL + C to quit)
Starting mock API Gateway service on http port 4566 ...
Starting mock CloudFormation service on http port 4566 ...
Starting mock CloudWatch service on http port 4566 ...
Starting mock DynamoDB service on http port 4566 ...
Starting mock DynamoDB Streams service on http port 4566 ...
Starting mock EC2 service on http port 4566 ...
Starting mock ES service on http port 4566 ...
Starting mock Firehose service on http port 4566 ...
Starting mock IAM service on http port 4566 ...
Starting mock STS service on http port 4566 ...
Starting mock Kinesis service on http port 4566 ...
Starting mock KMS service on http port 4566 ...
Starting mock Lambda service on http port 4566 ...
Starting mock CloudWatch Logs service on http port 4566 ...
Starting mock Redshift service on http port 4566 ...
Starting mock Route53 service on http port 4566 ...
Starting mock S3 service on http port 4566 ...
Starting mock Secrets Manager service on http port 4566 ...
Starting mock SES service on http port 4566 ...
Starting mock SNS service on http port 4566 ...
Starting mock SQS service on http port 4566 ...
Starting mock SSM service on http port 4566 ...
Starting mock Cloudwatch Events service on http port 4566 ...
Starting mock StepFunctions service on http port 4566 ...
2020-12-21T18:13:11:INFO:localstack.utils.analytics.profiler: Execution of "start_api_services" took 18576.085090637207ms
Ready.
```

**Comment by whummer** [comment]
That's fantastic @ryanone , great news! A PR would be much appreciated - we'll be happy to provide guidance and jump in where needed. Do you think you can create a PR with an initial version that we can iterate over together..? Many thanks for your help

**Comment by nsmithuk** [comment]
@ryanone @whummer I've recently added a hosted linux-arm64 build of local-kms.

The installer doesn't currently support any reference to cpu architecture, thus we've hosted binaries for localstack without said references for this reason. But if someone is looking at amending the installer, it would be good if it could changed such that it uses our standard naming, which would support this use-case.

For Linux we'd have:
```
https://s3-eu-west-2.amazonaws.com/local-kms/3/local-kms_linux-[amd64|arm64][-alpine].bin
```

For Mac we'd have:
```
https://s3-eu-west-2.amazonaws.com/local-kms/3/local-kms_darwin-[amd64|arm64].bin
```

So the full list of binaries available is:
```
local-kms_linux-amd64.bin
local-kms_linux-arm64.bin

local-kms_linux-amd64-alpine.bin
local-kms_linux-arm64-alpine.bin

local-kms_darwin-amd64.bin
local-kms_darwin-arm64.bin (experimental)
```

**Comment by ryanone** [comment]
Thanks for responding @whummer and your input @nsmithuk. I'll work on a PR and we can go from there.

**Comment by matthew-gill** [comment]
Hey all, any news on this one? Would love to give it a test and provide feedback :)

**Comment by mogggggg** [comment]
Hey @ryanone, did you have any luck with this? I'd be more than happy to help out

**Comment by chezsmithy** [comment]
@ryanone  also happy to help out getting this published.

**Comment by ryanone** [comment]
Ok, it's available at https://github.com/ryanone/localstack - it's likely even different from when I first got things to apparently work back then. But hopefully it'll help give a running start to those that can carry it over. Thanks for your patience with this - appreciate it.

**Comment by kipters** [comment]
Thanks! I'll try to get it running on my machine and update here if I'm successful

**Comment by whummer** [comment]
Awesome, @kipters - have you been able to get this working on your machine, with a Docker multi-arch build? Please keep us posted how it goes - happy to jump in an help out as well.. Thanks!

**Comment by kipters** [comment]
> Awesome, @kipters - have you been able to get this working on your machine, with a Docker multi-arch build? Please keep us posted how it goes - happy to jump in and help out as well.. Thanks!

I started working on it but no real progress due to lack of spare time, hopefully I might have some time to work on it this entire weekend.

**Comment by viswanath7** [comment]
Any update on this? I tried LocalStack version: 0.12.18 and still no ARM compatible image is available. When I run on a M1 Mac, I notice the following stack trace 

```
localstack_1  | 2021-10-16T18:47:58:INFO:bootstrap.py: Execution of "load_plugin_from_path" took 8105.87ms
localstack_1  | 2021-10-16T18:47:58:DEBUG:bootstrap.py: Plugin loading took 8.10982894897461 sec
localstack_1  | 2021-10-16T18:47:58:INFO:bootstrap.py: Execution of "load_plugins" took 8110.66ms
localstack_1  | 2021-10-16T18:47:58:DEBUG:localstack.services.install: getting kinesis-mock for linux x86_64
localstack_1  | 2021-10-16T18:47:58:DEBUG:localstack.services.install: kinesis-mock found at /opt/code/localstack/localstack/infra/kinesis-mock/kinesis-mock-linux-amd64-static
localstack_1  | 2021-10-16T18:47:58:DEBUG:localstack.utils.run: Executing command: cd "/opt/code/localstack/localstack/infra/dynamodb" && zip -u DynamoDBLocal.jar log4j2.xml || true
localstack_1  | Starting edge router (https port 4566)...
localstack_1  | Starting mock DynamoDB service on http port 4566 ...
localstack_1  | Starting mock DynamoDB Streams service on http port 4566 ...
localstack_1  | Starting mock Kinesis service on http port 4566 ...
localstack_1  | Starting mock SQS service on http port 4566 ...
localstack_1  | qemu: uncaught target signal 11 (Segmentation fault) - core dumped
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.utils.common: Starting download from https://github.com/localstack/localstack-artifacts/raw/master/local-certs/server.key to /tmp/localstack/server.test.pem (5159 bytes)
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.utils.common: Done downloading https://github.com/localstack/localstack-artifacts/raw/master/local-certs/server.key, response code 200, total bytes 7571
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.utils.run: Executing command: cd "/opt/code/localstack/localstack/infra/dynamodb" && zip -u DynamoDBLocal.jar log4j2.xml || true
localstack_1  | [2021-10-16 18:47:59 +0000] [51] [INFO] Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1  | 2021-10-16T18:47:59:INFO:hypercorn.error: Running on https://0.0.0.0:4566 (CTRL + C to quit)
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.utils.run: Executing command: cd /opt/code/localstack/localstack/infra/dynamodb/; java -Djava.library.path=./DynamoDBLocal_lib -Xmx256m -jar DynamoDBLocal.jar -port 54527 -inMemory
localstack_1  | 2021-10-16 18:47:59,563:API:  * Running on all addresses.
localstack_1  |    WARNING: This is a development server. Do not use it in a production deployment.
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.services.install: getting kinesis-mock for linux x86_64
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.services.install: kinesis-mock found at /opt/code/localstack/localstack/infra/kinesis-mock/kinesis-mock-linux-amd64-static
localstack_1  | 2021-10-16T18:47:59:DEBUG:localstack.utils.run: Executing command: KINESIS_MOCK_PLAIN_PORT=52959 SHARD_LIMIT=100 CREATE_STREAM_DURATION=500ms DELETE_STREAM_DURATION=500ms REGISTER_STREAM_CONSUMER_DURATION=500ms START_STREAM_ENCRYPTION_DURATION=500ms STOP_STREAM_ENCRYPTION_DURATION=500ms DEREGISTER_STREAM_CONSUMER_DURATION=500ms MERGE_SHARDS_DURATION=500ms SPLIT_SHARD_DURATION=500ms UPDATE_SHARD_COUNT_DURATION=500ms  LOG_LEVEL=INFO  /opt/code/localstack/localstack/infra/kinesis-mock/kinesis-mock-linux-amd64-static --gc=G1
localstack_1  | 2021-10-16T18:47:59:INFO:localstack.multiserver: Starting multi API server process on port 46937
localstack_1  | 2021-10-16T18:47:59:ERROR:localstack.services.kinesis.kinesis_starter: kinesis terminated with return code -11
```

**Comment by navels** [comment]
I saw that @whummer pushed `localstack/localstack:multi` recently, hoping that's a good sign!

**Comment by chezsmithy** [comment]
I have pushed a DRAFT PR with changes to allow building local arm versions of Dockerfile-base and Dockerfile. https://github.com/localstack/localstack/pull/4754

Next up, we need to think about adding in docker buildx build support into the makefile and cirlcleci configuration. This is going to take a bit of a change, because manifested builds can't be stored on the local amd64 machine executor in circle ci. Any good ideas how to allow me to test some CircleCI changes for you?

**Comment by chezsmithy** [comment]
@navels if you want to try things out. Pull down this branch, and setup the python environment, then run:

`make docker-build-mac-arm64`

That should output you a local arm ready build of the localstack/localstack:latest you can play with. Let me know how it goes.

**Comment by chezsmithy** [comment]
> Are you building in a Rosetta terminal? I'm getting errors building cryptography in a native terminal.
> 
> Got past it:
> 
> `LDFLAGS="-L$(brew --prefix openssl@1.1)/lib" CFLAGS="-I$(brew --prefix openssl@1.1)/include" make install`
> 
> Also had to install postgresql and jdk.

I'm not using Rosetta at all, I don't think. I defiantly have those ENVVARS from other Ruby work I've been doing in my .zshrc. So that makes complete sense.

**Comment by chezsmithy** [comment]
Current status. I've got both arm64 and amd64 builds running now locally and on CircleCI. I'm looking at tests now. 28 are failing for me locally, and I believe the majority are due to testing lambdaci images which aren't compatible with ARM, and I see the test runs failing in circleci on arm. I'd like to refactor the testing approach to spin up a docker image and then run the tests against it like make docker-test.

**Comment by navels** [comment]
I'm using the build now locally but our app only uses S3 so not a very comprehensive beta test :-)

**Comment by jbosse** [comment]
I was able to build the branch from @chezsmithy and FWIW my test suite that uses the container for local development and testing has about the same performance with the amd64 container as the arm64 container.

**Comment by thrau** [comment]
that's awesome news! yes i think we're almost over the line. @chezsmithy and @alexrashed have made amazing progress. there are still a few details that need ironing out, but i'm optimistic that we can get it released until next friday. /cc @whummer

**Comment by navels** [comment]
I just updated and am rebuilding via `make docker-build-multi-platform` and it's been sitting on this step for over an hour:

```
 => [linux/amd64 java-builder 3/3] RUN jlink --add-modules java.base,java.desktop,java.instrument,java.management,java.naming,java.scripting,java.sql,java.xml,jdk.compiler,jdk.unsupported,jdk.crypto.cr
```

Is this expected?

**Comment by chezsmithy** [comment]
@navels try building with `make docker-build-linux-arm64`, if you just want an arm image you can load to docker on your M1 Mac.

**Comment by navels** [comment]
> @navels try building with `make docker-build-linux-arm64`, if you just want an arm image you can load to docker on your M1 Mac.

Thanks, I missed that target . . .  works fine!

**Comment by thrau** [comment]
**TL;DR:** we're planing to release an initial version of multi-platform support in 0.13. next week, where amd64 will be fully supported, and there will be experimental support for aarch64

---

Hi Everyone, here's a quick update from the LocalStack team on the state of things:

I want to highlight once more @chezsmithy on the community side of things, who's laid fantastic groundwork for the multi-platform build in #4754! we've been working internally the past week to get everything over the line. it's been quite an involved process, as there are lots of moving parts, as well as challenges introduced by the changes we've had to make.

Here's a summary:

* we've switched completely from an alpine-based Dockerfile to debian buster, both for amd64 and aarch64
* this required deep-reaching changes and fixes in both the community and our pro code base (which have naturally not been visible here)
* our goal was to have a green CI run of LocalStack Pro using Debian-based amd64 image. This was successful today!
* there are still several workarounds and limitations for aarch64, which we are currently trying to tally, and there will be a detailed list of limitations in the release notes. as of now, it looks like KMS, ECR, and kinesisanalytics will not work in aarch64 in the first release.
* we didn't quite get the release done for today, but there will be a release of LocalStack 0.13 next week, with the working debian docker image for amd64, and limited support for the experimental aarch64 version.
* it will be an ongoing effort over a few weeks following the release to iron out the bugs and limitations for aarch64, until we can confidently support it. 

we're excited to get aarch64 out there for everyone to test in 0.13, but we don't expect everything to work immediately, so please be patient with us. you would help us immensely with bug reports :-)

that's it for now, i'll notify everyone once the release is out there next week.

**Comment by alexrashed** [comment]
Hi everyone!
With version 0.13.0, the multi-platform docker image has officially been released! 🥳 
As @thrau mentioned, initially the arm64 image is still experimental.
You can find information on the currently known limitations in #4921.
When you experience any problems (arm64 specific or not), please help us getting aware of it by creating a bug report.
